prestoadmin/server.py

```diff
@@ -82,11 +86,191 @@ def new_sysnode_processor(node_info_rows):
+            download_buffer = url_response.read(block_size)
+            if not download_buffer:
+                break
+            bytes_read += len(download_buffer)
+            local_file.write(download_buffer)
+            if content_length:
+                _print_download_status(bytes_read, content_length)
+        print "Downloaded %d bytes." % bytes_read
+
+    print 'Rpm downloaded to: %s' % download_file_path
+    return download_file_path
+
+
+def _find_or_download_rpm(url, version=None):
+    """
+    Args:
+        url:      The url of the presto rpm to be downloaded.
+        version:  An optional version number.
+                  If the server doesn't respond with the file name that is being
+                  requested, this allows the downloaded file to have the correct
+                  version attached to its name (presto-server-rpm-'version'.rpm)
+                  rather than the default name
+
+    If downloading the presto rpm at the given url would overwrite an existing rpm,
+    this function returns the path to the existing rpm. However, if the rpm that
+    would be downloaded takes the default rpm name, it will overwrite the existing
+    rpm because there is no way to know if the default rpm name is of the same version
+    as the requested rpm.
+
+    Returns:
+        Upon success, the path to the downloaded or found presto rpm
+        Upon failure, None
+    """
+    try:
+        url_response = urllib2.urlopen(url)
+    except ValueError:
+        return None
+
+    if not _check_good_response_status(url_response.getcode()):
+        return None
+
+    download_file_name = _get_download_file_name(url_response, version)
+    download_file_path = os.path.join(DOWNLOAD_DIRECTORY, download_file_name)
+
+    downloaded_rpm_path = _find_downloaded_rpm(download_file_name)
+    if downloaded_rpm_path and download_file_name != DEFAULT_RPM_NAME:
+        return downloaded_rpm_path
+
+    return _download_rpm(url_response, download_file_path)
+
+
+def _find_or_download_rpm_version_facebook(rpm_version):
+    rpm_version_fb = rpm_version[:-1]
+    download_url = 'http://search.maven.org/remotecontent?filepath=com/facebook/presto/' \
+                   'presto-server-rpm/' + rpm_version_fb + '/presto-server-rpm-' + \
+                   rpm_version_fb + '.rpm'
+    return _find_or_download_rpm(download_url, rpm_version_fb)
+
+
+def _find_or_download_rpm_version_teradata(rpm_version):
+    abort('Download for teradata presto rpms is not currently supported.\n'
+          'Try again after downloading the rpm and specifying the local path to the rpm')
+
+
+def _find_or_download_rpm_version(rpm_version):
+    """
+    Attempt to find or download the given rpm version.
+
+    Default to using Facebook rpm if 'f'aceboook and 't'eradata
+    are not specified in rpm_version.
+    """
+    if not _check_if_valid_version(rpm_version):
+        return None
+
+    if rpm_version[-1] != 'f' and rpm_version[-1] != 't':
```

Comment:
I don't think we guarantee that 't' will be the last component in the version. 0.148-t.1.1 is a version number that's been mentioned in #git-workflow.